granulate_utils/metadata/databricks_client.py

```diff
@@ -115,109 +146,107 @@ def _discover_apps_url(self) -> bool:
             return True
 
-    def _spark_apps_json(self) -> Any:
+    def _fetch_spark_apps(self) -> list:
         assert self._apps_url, "SparkUI apps url was not discovered"
-        try:
-            response = self._request_get(self._apps_url)
-        except requests.exceptions.RequestException:
-            # Request might fail in cases where the cluster is still initializing, retrying.
-            return None
+        response = self._request_get(self._apps_url)
         try:
             apps = response.json()
         except Exception as e:
             if "Spark is starting up. Please wait a while until it's ready" in response.text:
                 # Spark is still initializing, retrying.
                 # https://github.com/apache/spark/blob/38c41c/core/src/main/scala/org/apache/spark/ui/SparkUI.scala#L64
-                return None
+                raise SparkNotReadyException("Spark is still initializing") from e
             else:
-                raise DatabricksJobNameDiscoverException(
-                    f"Failed to parse apps url response, query {response.text=}"
-                ) from e
+                raise SparkAPIException(f"Failed to parse apps url response, query {response.text=}") from e
         return apps
 
-    def _spark_app_env_json(self, app_id: str) -> Any:
+    def _fetch_spark_app_environment(self, app_id: str) -> Dict[str, Any]:
         assert self._apps_url is not None, "SparkUI apps url was not discovered"
         env_url = f"{self._apps_url}/{app_id}/environment"
-        try:
-            response = self._request_get(env_url)
-        except Exception as e:
-            # No reason for any exception, `environment` uri should be accessible if we have running apps.
-            raise DatabricksJobNameDiscoverException(f"Environment request failed {env_url=}") from e
-        try:
-            env = response.json()
-        except Exception as e:
-            raise DatabricksJobNameDiscoverException(f"Environment request failed {response.text=}") from e
-        return env
+        response = self._request_get(env_url)
+        return response.json()
 
-    def _cluster_all_tags_metadata(self) -> Optional[Dict[str, str]]:
+    def _fetch_cluster_all_tags(self) -> Dict[str, str]:
         """
-        Returns `includes spark.databricks.clusterUsageTags.clusterAllTags` tags as `Dict`.
+        Returns `includes spark.databricks.clusterUsageTags.clusterAllTags` tags as `dict`.
         In any case this function returns `None`, a retry is required.
         """
         if not os.path.isfile(DATABRICKS_METRICS_PROP_PATH):
+            self.logger.warning(f"{DATABRICKS_METRICS_PROP_PATH} was not found.")
             # We want to retry in case the cluster is still initializing, and the file is not yet deployed.
-            return None
+            raise SparkAppsURLDiscoveryException(f"{DATABRICKS_METRICS_PROP_PATH} was not found.")
```

Comment:
Logging messages before exceptions looks redundant. All the exceptions are always logged by the caller anyway afaict. If the log level is important, then we can add a value to the exception classes that says which level they should be logged at.

Reply:
I agree but I'm afraid of a situation where I'm unaware of all the callers, this is why I also added an argument indicating whether to raise or not.  
It's either:
a. We drop all the log messages
b. We'll log only if not raise arg is opted in